src/js/components/tasks/retrieveDataset/index.tsx

```diff
@@ -0,0 +1,312 @@
+  fetchingDatasets: boolean;
+  isOpen: boolean;
+  closeModal: () => void;
+}
+
+export interface DatasetCommit {
+  dataset_name: string;
+  dataset_definition: Changeset;
+  commit_message: string;
+}
+
+export interface BooleanObject {
+  [key: string]: boolean;
+}
+
+const RetrieveDatasetModal = ({
+  orgId,
+  datasets,
+  datasetErrors,
+  schema,
+  fetchingDatasets,
+  isOpen,
+  closeModal,
+}: Props) => {
+  const { t } = useTranslation();
+  const dispatch = useDispatch<ThunkDispatch>();
+  const [retrievingDataset, setRetrievingDataset] = useState(false);
+  const [pageIndex, setPageIndex] = useState(0);
+  const [hasError, setHasError] = useState(false);
+  const [outdatedChangeset, setOutdatedChangeset] = useState<Changeset>({});
+  const isMounted = useIsMounted();
+
+  const datasetNames = Object.keys(datasets);
+  const orderedDatasetNames = [
+    ...datasetNames.filter((name) => name.toLowerCase() === 'default'),
+    ...sortBy(
+      datasetNames.filter((name) => name.toLowerCase() !== 'default'),
+      toLower,
+    ),
+  ];
+
+  const nextPage = () => {
+    setPageIndex(pageIndex + 1);
+  };
+
+  const prevPage = () => {
+    setPageIndex(Math.max(pageIndex - 1, 0));
+  };
+
+  const handleSuccess = () => {
+    /* istanbul ignore else */
+    if (isMounted.current) {
+      setRetrievingDataset(false);
+      closeModal();
+      setPageIndex(0);
+    }
+  };
+
+  // eslint-disable-next-line handle-callback-err
+  const handleError = (
+    err: ApiError,
+    fieldErrors: { [key: string]: string },
+  ) => {
+    /* istanbul ignore else */
+    if (isMounted.current) {
+      setRetrievingDataset(false);
+      if (fieldErrors.dataset_name) {
+        setPageIndex(0);
+      } else if (fieldErrors.dataset_definition) {
+        setPageIndex(1);
+      } else if (fieldErrors.commit_message) {
+        setPageIndex(2);
+      }
+    }
+  };
+
+  const {
+    inputs,
+    errors,
+    handleInputChange,
+    setInputs,
+    handleSubmit,
+    resetForm,
+  } = useForm({
+    fields: {
+      dataset_name: '',
+      dataset_definition: {},
+      commit_message: '',
+    } as DatasetCommit,
+    objectType: OBJECT_TYPES.COMMIT_DATASET,
+    url: window.api_urls.scratch_org_commit_dataset(orgId),
+    onSuccess: handleSuccess,
+    onError: handleError,
+    shouldSubscribeToObject: false,
+  });
+
+  // When datasets change, update default selection
+  const selectedDatasetRef = useRef(inputs.dataset_name);
+  useEffect(() => {
+    const selectedDataset = inputs.dataset_name;
+    const prevValue = selectedDatasetRef.current;
+    if (selectedDataset !== prevValue) {
+      const def = Object.assign({}, datasets[inputs.dataset_name] ?? {});
+      // @@@ Adding fake outdated fields
+      // if (datasets[inputs.dataset_name]) {
+      //   def.foo = ['bar', 'buz'];
+      //   for (let it = 1; it < 20; it = it + 1) {
+      //     def[`${it}-object`] = [`${it}-field`];
+      //   }
+      // }
```

Comment:
@mirisuzanne You can uncomment these lines to fake the case where old schema items will be removed during the capture process.